README.md
# PHP JSON-G

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-github-actions]][link-github-actions]
[![Style CI][ico-styleci]][link-styleci]
[![Total Downloads][ico-downloads]][link-downloads]
[![Buy us a tree][ico-treeware-gifting]][link-treeware-gifting]

A PHP implementation of [JSON-G](https://github.com/Roadcrosser/JSON-G).

## Install

Via Composer

```bash
$ composer require pxgamer/json-g
```

## Usage

Use the `JsonG::toImageBlob()` or `JsonG::toJson()` to convert between a blob and JSON-G.

#### Examples

**Converting an image via POST**

```php
use RaidAndFade\JsonG\JsonG;

$base64 = base64_decode($_POST['in']);

$image = new Imagick();
$image->readimageblob($base64);

$json = JsonG::toJson($image);

header("Content-Type: text/json");
echo $json;
```

**Converting a JSON-G string to image via POST**

```php
use RaidAndFade\JsonG\JsonG;

$jsonArray = json_decode($_POST['in'], true);

$image = JsonG::toImageBlob($jsonArray);

header("Content-Type: image/png");
echo $image;
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

```shell
composer test
```

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email security@voke.dev instead of using the issue tracker.

## Credits

- [Owen Voke][link-author]
- [RaidAndFade](https://git.gocode.it/RaidAndFade/PHP_json-g)
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Treeware

You're free to use this package, but if it makes it to your production environment you are required to buy the world a tree.

It’s now common knowledge that one of the best tools to tackle the climate crisis and keep our temperatures from rising above 1.5C is to plant trees. If you support this package and contribute to the Treeware forest you’ll be creating employment for local families and restoring wildlife habitats.

You can buy trees [here][link-treeware-gifting].

Read more about Treeware at [treeware.earth][link-treeware].

[ico-version]: https://img.shields.io/packagist/v/pxgamer/json-g.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-github-actions]: https://img.shields.io/github/workflow/status/pxgamer/php-json-g/Tests.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/107679531/shield
[ico-downloads]: https://img.shields.io/packagist/dt/pxgamer/json-g.svg?style=flat-square
[ico-treeware-gifting]: https://img.shields.io/badge/Treeware-%F0%9F%8C%B3-lightgreen?style=flat-square

[link-packagist]: https://packagist.org/packages/pxgamer/json-g
[link-github-actions]: https://github.com/pxgamer/php-json-g/actions
[link-styleci]: https://styleci.io/repos/107679531
[link-downloads]: https://packagist.org/packages/pxgamer/json-g
[link-treeware]: https://treeware.earth
[link-treeware-gifting]: https://ecologi.com/owenvoke?gift-trees
[link-author]: https://github.com/owenvoke
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `php-json-g` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](https://keepachangelog.com) principles.

## Unreleased

## [v2.2.0] - 2022-02-14

### Added
- Add array shaping for static analysis
- Add static analysis testing in CI

### Changed
- Add strict return types for methods
- Replace Travis CI with GitHub Actions

## [v2.1.0] - 2019-03-11

### Added
- Add support for PHPUnit 8 ([#2](https://github.com/pxgamer/php-json-g/pull/2))
- Add missing extension references to the Composer JSON ([b2ee756b](https://github.com/pxgamer/php-json-g/commit/b2ee756b126ed910151e2fe5d25244b89e126cc0))

### Changed
- Update the directory structure formatting ([70dfbf55](https://github.com/pxgamer/php-json-g/commit/70dfbf5536e7d45816c2bbd9c7505676321ece7a))
- Move test resources to the `tests` directory ([dcf5d675](https://github.com/pxgamer/php-json-g/commit/dcf5d675eb490f24b4d39b777099a72c55936826))

### Deprecated
- Remove support for PHP 7.1 ([19d12ecb](https://github.com/pxgamer/php-json-g/commit/19d12ecba15ef32be52c5ee552d28fbec6d99218))

## [v2.0.3] - 2017-12-06

### Added
- Bumped `squizlabs/php_codesniffer` dependency to `^3.1`
- Updated PhpDoc styling
- Added a `ColoursTest` class to get 100% in the Colours coverage

## [v2.0.2] - 2017-12-06

### Fixed
- Changed to legal name in the LICENSE

## [v2.0.1] - 2017-11-20

### Fixed
- Corrected the LICENSE file

## [v2.0.0] - 2017-11-08

### Added
- Added new README style
- Added namespacing for tests
- Added missing OS files

### Fixed
- Fixed the styling using StyleCI and PHP_CodeSniffer
- Fixed TravisCI testing using PECL/Imagick

## v1.0.0 - 2017-10-20

### Added
- Added 2 classes for managing JSON-G image conversion
- Added Composer support
- Added unit testing
- Added namespacing

[v2.2.0]: https://github.com/pxgamer/php-json-g/compare/v2.1.0...v2.2.0
[v2.1.0]: https://github.com/pxgamer/php-json-g/compare/v2.0.3...v2.1.0
[v2.0.3]: https://github.com/pxgamer/php-json-g/compare/v2.0.2...v2.0.3
[v2.0.2]: https://github.com/pxgamer/php-json-g/compare/v2.0.1...v2.0.2
[v2.0.1]: https://github.com/pxgamer/php-json-g/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/pxgamer/php-json-g/compare/v1.0.0...v2.0.0


tests/ColoursTest.php
<?php

use RaidAndFade\JsonG\Colours;

it('can create an ImagickPixel with RGB', function () {
    $pixel = Colours::toImgPixel([
        'r' => 1,
        'g' => 1,
        'b' => 1,
    ]);

    expect($pixel)->toBeInstanceOf(ImagickPixel::class);
});

it('can create an ImagickPixel with RGBA', function () {
    $pixel = Colours::toImgPixel([
        'r' => 1,
        'g' => 1,
        'b' => 1,
        'a' => 1,
    ]);

    expect($pixel)->toBeInstanceOf(ImagickPixel::class);
});


tests/JsonGToImageTest.php
<?php

use RaidAndFade\JsonG\JsonG;

it('can convert a JSON-G string to an image blob', function () {
    $testData = file_get_contents(__DIR__.'/resources/demo.jsng');

    $jsonArray = json_decode($testData, true);

    $image = JsonG::toImageBlob($jsonArray);

    expect($image)->toBeString();
});


tests/JsonGFromImageTest.php
<?php

use RaidAndFade\JsonG\JsonG;

it('can convert an image blob to a JSON-G string', function () {
    $testData = fopen(__DIR__.'/resources/demo.png', 'rb');

    $image = new Imagick();
    $image->readImageFile($testData);

    $json = JsonG::toJson($image);

    expect($json)->toBeJson();
    $this->assertJsonStringEqualsJsonFile(__DIR__.'/resources/demo.jsng', $json);
});


src/JsonG.php
<?php

namespace RaidAndFade\JsonG;

use Imagick;
use ImagickDraw;
use ImagickDrawException;
use ImagickException;
use ImagickPixel;
use ImagickPixelException;
use JsonException;

class JsonG
{
    /**
     * Convert a JSON-G array instance to a Imagick image blob.
     *
     * @param  array{size: array{width: int, height: int}, layers: array<string|int, array{default_color?: array{red: int, green: int, blue: int, alpha?: int}, default_colour?: array{red: int, green: int, blue: int, alpha?: int}, pixels: array<string|int, array{color?: array{red: int, green: int, blue: int, alpha?: int}, colour?: array{red: int, green: int, blue: int, alpha?: int}, position: array{x: int, y: int}}>}>}  $jsonArray
     * @return string
     *
     * @throws ImagickException|ImagickDrawException
     */
    public static function toImageBlob(array $jsonArray): string
    {
        $image = new Imagick();
        $image->newImage($jsonArray['size']['width'], $jsonArray['size']['height'], new ImagickPixel('none'));
        $image->setImageFormat('png');

        foreach ($jsonArray['layers'] as $layerId => $layer) {
            $layerDrawing = new ImagickDraw();

            if (isset($layer['default_colour']) && ! isset($layer['default_color'])) {
                $layer['default_color'] = $layer['default_colour'];
            }

            assert(isset($layer['default_color']), "A default color was not set on layer `{$layerId}`");

            $layerDrawing->setFillColor(Colours::toImgPixel(Colours::fullToShort($layer['default_color'])));
            $layerDrawing->setStrokeWidth(0);
            $layerDrawing->rectangle(0, 0, $jsonArray['size']['width'], $jsonArray['size']['height']);
            $layerDrawing->setStrokeWidth(1);

            foreach ($layer['pixels'] as $pixelId => $pixel) {
                if (isset($pixel['colour']) && ! isset($pixel['color'])) {
                    $pixel['color'] = $pixel['colour'];
                }

                assert(isset($pixel['color']), "A color was not set on layer `{$layerId}` and pixel `{$pixelId}`");

                $layerDrawing->setFillColor(Colours::toImgPixel(Colours::fullToShort($pixel['color'])));
                $layerDrawing->point($pixel['position']['x'], $pixel['position']['y']);
            }

            $image->drawImage($layerDrawing);
        }

        return $image->getImageBlob();
    }

    /**
     * Converts an Imagick image instance to a JSON-G string.
     *
     * @param  Imagick  $image
     * @return string
     *
     * @throws ImagickPixelException|ImagickException
     * @throws JsonException
     */
    public static function toJson(Imagick $image): string
    {
        /** @var array{color?: int, colour?: int, position: array{x: int, y: int}}  $pixels */
        $pixels = [];
        $colors = [];
        $res = [
            'version' => '1.0',
            'transparency' => true,
            'layers' => [],
            'size' => [
                'width' => $image->getImageWidth(),
                'height' => $image->getImageHeight(),
            ],
        ];

        for ($x = 0; $x <= $image->getImageWidth(); $x++) {
            $pixels[$x] = [];
            for ($y = 0; $y <= $image->getImageHeight(); $y++) {
                $pixels[$x][$y] = $image->getImagePixelColor($x, $y)->getColor();
            }
        }

        foreach ($pixels as $cols) {
            foreach ($cols as $pixel) {
                $col = Colours::toInt($pixel);
                if (array_key_exists($col, $colors)) {
                    $colors[$col]++;
                } else {
                    $colors[$col] = 1;
                }
            }
        }

        $def = array_search(max($colors), $colors, true);
        $l = [
            'pixels' => [],
            'default_color' => Colours::shortToFull(Colours::fromInt($def)), // @phpstan-ignore-line
        ];

        foreach ($pixels as $x => $cols) {
            foreach ($cols as $y => $pixel) {
                $col = ($pixel['a'] << 24) + ($pixel['r'] << 16) + ($pixel['g'] << 8) + $pixel['b'];
                if ($col !== $def) {
                    $l['pixels'][] = [
                        'position' => compact('x', 'y'),
                        'color' => Colours::shortToFull($pixel),
                    ];
                }
            }
        }

        $res['layers'][0] = $l;

        return json_encode($res, JSON_THROW_ON_ERROR);
    }
}


src/Colours.php
<?php

namespace RaidAndFade\JsonG;

use ImagickPixel;

class Colours
{
    /**
     * Convert a colour array to an integer sequence.
     *
     * @param  array{r: int, g: int, b: int, a: int}  $colour
     */
    public static function toInt(array $colour): int
    {
        return ($colour['a'] << 24) + ($colour['r'] << 16) + ($colour['g'] << 8) + $colour['b'];
    }

    /**
     * Convert an integer sequence to an array.
     *
     * @return array{r: int, g: int, b: int, a: int}
     */
    public static function fromInt(int $int): array
    {
        return [
            'r' => 0xFF & $int >> 16,
            'g' => 0xFF & $int >> 8,
            'b' => 0xFF & $int,
            'a' => 0xFF & $int >> 24,
        ];
    }

    /**
     * Convert a colour array to an ImagickPixel instance.
     *
     * @param  array{r: int, g: int, b: int, a?: int}  $colour
     */
    public static function toImgPixel(array $colour): ImagickPixel
    {
        if (isset($colour['a'])) {
            $col = sprintf('rgba(%s,%s,%s,%s)', $colour['r'], $colour['g'], $colour['b'], $colour['a']);
        } else {
            $col = sprintf('rgb(%s,%s,%s)', $colour['r'], $colour['g'], $colour['b']);
        }

        return new ImagickPixel($col);
    }

    /**
     * Convert a short (RGBA) array to a full-word array (red, green, blue, alpha).
     *
     * @param  array{r: int, g: int, b: int, a?: int}  $colour
     * @return array{red: int, green: int, blue: int, alpha?: int}
     */
    public static function shortToFull(array $colour): array
    {
        $arr = [
            'red' => $colour['r'],
            'green' => $colour['g'],
            'blue' => $colour['b'],
        ];

        if (isset($colour['a'])) {
            $arr['alpha'] = $colour['a'];
        }

        return $arr;
    }

    /**
     * Convert a full-word (red, green, blue, alpha) array to a  short (RGBA) array.
     *
     * @param  array{red: int, green: int, blue: int, alpha?: int}  $colour
     * @return array{r: int, g: int, b: int, a?: int}
     */
    public static function fullToShort(array $colour): array
    {
        $arr = [
            'r' => $colour['red'],
            'g' => $colour['green'],
            'b' => $colour['blue'],
        ];

        if (isset($colour['alpha'])) {
            $arr['a'] = $colour['alpha'];
        }

        return $arr;
    }
}


jsong.php
<?php
function isJson($string) {
  json_decode($string);
  return (json_last_error() == JSON_ERROR_NONE);
}
function result($msg="",$code=200){
    http_response_code($code);
    die($msg);
}
function colToInt($c){
  return ($c['a']<<24) + ($c['r']<<16) + ($c['g']<<8) + $c['b'];
}
function intToCol($i){
  return array('r' => 0xff & $i >> 16 ,
               'g' => 0xff & $i >>  8 ,
               'b' => 0xff & $i       ,
               'a' => 0xff & $i >> 24
        );
}
function colToImgPixel($c){
  if(isset($c['a']))
    $col = "rgba(".$c['r'].",".$c['g'].",".$c['b'].",".$c['a'].")";
  else
    $col = "rgb(".$c['r'].",".$c['g'].",".$c['b'].")";
  return new ImagickPixel($col);
}
function shortToFullCols($c){
  $arr = array("red"=>$c['r'],"blue"=>$c['b'],"green"=>$c['g']);
  if(isset($c['a']))    $arr['alpha']=$c['a'];
  return $arr;
}
function fullToShortCols($c){
  $arr = array("r"=>$c['red'],"b"=>$c['blue'],"g"=>$c['green']);
  if(isset($c['alpha']))    $arr['a']=$c['alpha'];
  return $arr;
}
//Where the cancer begins
function jsongToIMG($js){
  $im = new Imagick();
  $im -> newImage($js['size']['width'],$js['size']['height'],new ImagickPixel('none'));
  $im -> setImageFormat('png');
  foreach ($js['layers'] as $layer) {
    $lid = new ImagickDraw();
    if(isset($layer['default_colour'])&&!isset($layer['default_color'])) $pixel['default_color'] = $layer['default_colour'];
    $lid -> setFillColor(colToImgPixel(fullToShortCols($layer['default_color'])));
    $lid -> setStrokeWidth(0);
    $lid -> rectangle(0,0,$js['size']['width'],$js['size']['height']);
    $lid -> setStrokeWidth(1);
    foreach($layer['pixels'] as $pixel){
      if(isset($pixel['colour'])&&!isset($pixel['color'])) $pixel['color'] = $pixel['colour'];
      $lid -> setFillColor(colToImgPixel(fullToShortCols($pixel['color'])));
      $lid -> point($pixel['position']['x'],$pixel['position']['y']);
    }
    $im -> drawImage($lid);
  }
  return $im -> getImageBlob();
}
function imgToJsong($im){
  $res = array();
  $res['version'] = "1.0";
  $res['transparency'] = true;
  $res['size'] = array();
  $res['size']['width'] = $im->getImageWidth();
  $res['size']['height'] = $im->getImageHeight();
  $res['layers'] = array();
  $pixels = array();
  for($x=0;$x<=$im->getImageWidth();$x++){
    $pixels[$x] = array();
    for($y=0;$y<=$im->getImageHeight();$y++){
      $pixels[$x][$y] = $im->getImagePixelColor($x,$y)->getColor();
    }
  }
  $colors = array();
  foreach ($pixels as $cols) {
    foreach ($cols as $pixel) {
      $col = colToInt($pixel);
      if(array_key_exists($col,$colors))
        $colors[$col]++;
      else {
        $colors[$col]=1;
      }
    }
  }
  $def = array_search(max($colors),$colors);
  $l = array();
  $l['default_color'] = shortToFullCols(intToCol($def));
  $l['pixels'] = array();
  foreach ($pixels as $x=>$cols) {
    foreach ($cols as $y=>$pixel) {
      $col = ($pixel['a']<<24) + ($pixel['r']<<16) + ($pixel['g']<<8) + $pixel['b'];
      if($col != $def){
        array_push(
          $l['pixels'],
          array(
            "position"=>  array(
              "x" =>  $x,
              "y" =>  $y
            ),
            "color"   =>  shortToFullCols($pixel)
          )
        );
      }
    }
  }
  $res['layers'][0] = $l;
  return json_encode($res);
}
//Where the cancer ends

if($_SERVER['REQUEST_METHOD']!="POST"){
  result("Method not allowed.",405);
}
if(!isset($_POST['in'])){
  result("Input is invalid.",400);
}
if(isJson($_POST['in'])){
  die($_POST['in']);
  $im = jsongToIMG(json_decode($_POST['in'],true));
  header("Content-Type: image/png");
  echo $im;
  result();
}
$b64 = base64_decode($_POST['in']);
if($b64){
  $im = new Imagick();
  $im -> readimageblob($b64);
  $js = IMGToJsong($im);
  result($js);
}
result("No valid input detected.",400);
//blah blah if file was uploaded directly fuck that guy cus postman doesnt handle file uploads properly and im too lazy to figure it out
?>
